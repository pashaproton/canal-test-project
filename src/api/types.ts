import { SORT_BY } from '../types';

export interface RequestParams {
  page: number
  adult?: boolean
  query?: string
  api_key: string
  sort_by?: SORT_BY
  language?: string
}

export interface DefaultResponse<T> {
  page: number
  results: T[]
  total_pages: number
  total_results: number
}

export interface GenericResponseResult {
  id: number
  overview: string
  genre_ids: number[]
  popularity: number
  vote_count: number
  poster_path?: string
  vote_average: number
  backdrop_path?: string
  original_language: string
}

export interface ResponseTV extends GenericResponseResult {
  name: string
  original_name: string
  first_air_date: string
  origin_country: string[]
}

export interface ResponseMovie extends GenericResponseResult {
  adult: boolean
  title: string
  video: boolean
  release_date: string
  original_title: string
}
