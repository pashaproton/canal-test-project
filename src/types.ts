import { Dispatch, SetStateAction } from 'react';

export enum SORT_BY {
  POPULARITY_ASC = 'popularity.asc',
  POPULARITY_DESC = 'popularity.desc',
  RELEASE_DATE_ASC = 'release_date.asc',
  RELEASE_DATE_DESC = 'release_date.desc',
  REVENUE_ASC = 'revenue.asc',
  REVENUE_DESC = 'revenue.desc',
  ORIGINAL_TITLE_ASC = 'original_title.asc',
  ORIGINAL_TITLE_DESC = 'original_title.desc',
  VOTE_AVERAGE_ASC = 'vote_average.asc',
  VOTE_AVERAGE_DESC = 'vote_average.desc',
  VOTE_COUNT_ASC = 'vote_count.asc',
  VOTE_COUNT_DESC = 'vote_count.desc',
  PRIMARY_RELEASE_DATE_ASC = 'primary_release_date.asc',
  PRIMARY_RELEASE_DATE_DESC = 'primary_release_date.desc'
}

export interface Movie {
  id: number
  date: string
  title: string
  voteCount: number
  imagePath?: string
  popularity: number
  voteAverage: number
  description: string
}

export interface Context {
  adult: boolean
  search: string
  sortBy: SORT_BY
  setAdult: Dispatch<SetStateAction<boolean>>
  setSearch: Dispatch<SetStateAction<string>>
  setSortBy: Dispatch<SetStateAction<SORT_BY>>
}