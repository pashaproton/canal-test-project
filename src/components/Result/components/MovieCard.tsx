import React from 'react';

import Card from '@mui/material/Card';
import Grid from '@mui/material/Grid';
import CardMedia from '@mui/material/CardMedia';
import Typography from '@mui/material/Typography';
import CardContent from '@mui/material/CardContent';

import { Movie } from '../../../types';
import { getImageUrl } from '../../../helpers';

interface MovieCardProps {
  data: Movie
}

const MovieCard: React.FC<MovieCardProps> = ({
  data
}) => {
  return (
    <Card sx={{ maxWidth: 345 }}>
      {data.imagePath && (
        <CardMedia
          component="img"
          height="140"
          image={getImageUrl(data.imagePath)}
          alt={data.title}
          data-testid="movie-card-image"
        />
      )}
      <CardContent>
        <Typography data-testid="movie-card-title" gutterBottom variant="h5" component="div">
          {data.title}
        </Typography>
        <Typography data-testid="movie-card-description" variant="body2" color="text.secondary">
          {data.description}
        </Typography>
        <hr />
        <Typography data-testid="movie-card-popularity" variant="body2" color="text.secondary">
          Popularity: {data.popularity}
        </Typography>
        <Grid container justifyContent="space-between">
          <Typography data-testid="movie-card-rate" variant="body2" color="text.secondary">
            {data.voteAverage} ({data.voteCount})
          </Typography>
          <Typography data-testid="movie-card-date" variant="body2" color="text.secondary" align="right">
            {data.date}
          </Typography>
        </Grid>
      </CardContent>
    </Card>
  );
};

export default React.memo(MovieCard);
