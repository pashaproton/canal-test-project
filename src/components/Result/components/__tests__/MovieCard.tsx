import React from 'react';
import { render, screen } from '@testing-library/react';
import { Movie } from '../../../../types';
import MovieCard from '../MovieCard';

describe('MovieCard', () => {
  const data: Movie = {
    id: 123,
    date: '05-05-2022',
    title: 'awesome movie',
    popularity: 9999,
    imagePath: '/path-to-image',
    voteCount: 777,
    voteAverage: 9.9,
    description: 'some overview'
  };

  it('should render the component with data', () => {
    render(<MovieCard data={data} />)
    expect(screen.getByTestId('movie-card-date')).toHaveTextContent('05-05-2022');
    expect(screen.getByTestId('movie-card-image'))
      .toHaveAttribute('src', 'https://image.tmdb.org/t/p/w300/path-to-image');
    expect(screen.getByTestId('movie-card-rate')).toHaveTextContent('9.9 (777)');
    expect(screen.getByTestId('movie-card-title')).toHaveTextContent('awesome movie');
    expect(screen.getByTestId('movie-card-popularity')).toHaveTextContent('9999');
    expect(screen.getByTestId('movie-card-description')).toHaveTextContent('some overview');
  });

  it('should render the component with data and compere it to snapshot', () => {
    const result = render(<MovieCard data={data} />);
    expect(result).toMatchSnapshot();
  });
});