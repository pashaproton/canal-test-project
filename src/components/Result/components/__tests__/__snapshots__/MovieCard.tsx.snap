// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieCard should render the component with data and compere it to snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-1ri6ub7-MuiPaper-root-MuiCard-root"
      >
        <img
          alt="awesome movie"
          class="MuiCardMedia-root MuiCardMedia-media MuiCardMedia-img css-o69gx8-MuiCardMedia-root"
          data-testid="movie-card-image"
          height="140"
          src="https://image.tmdb.org/t/p/w300/path-to-image"
        />
        <div
          class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
        >
          <div
            class="MuiTypography-root MuiTypography-h5 MuiTypography-gutterBottom css-h93ljk-MuiTypography-root"
            data-testid="movie-card-title"
          >
            awesome movie
          </div>
          <p
            class="MuiTypography-root MuiTypography-body2 css-r40f8v-MuiTypography-root"
            data-testid="movie-card-description"
          >
            some overview
          </p>
          <hr />
          <p
            class="MuiTypography-root MuiTypography-body2 css-r40f8v-MuiTypography-root"
            data-testid="movie-card-popularity"
          >
            Popularity: 
            9999
          </p>
          <div
            class="MuiGrid-root MuiGrid-container css-1xzqqmo-MuiGrid-root"
          >
            <p
              class="MuiTypography-root MuiTypography-body2 css-r40f8v-MuiTypography-root"
              data-testid="movie-card-rate"
            >
              9.9
               (
              777
              )
            </p>
            <p
              class="MuiTypography-root MuiTypography-body2 MuiTypography-alignRight css-15l36n3-MuiTypography-root"
              data-testid="movie-card-date"
            >
              05-05-2022
            </p>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-1ri6ub7-MuiPaper-root-MuiCard-root"
    >
      <img
        alt="awesome movie"
        class="MuiCardMedia-root MuiCardMedia-media MuiCardMedia-img css-o69gx8-MuiCardMedia-root"
        data-testid="movie-card-image"
        height="140"
        src="https://image.tmdb.org/t/p/w300/path-to-image"
      />
      <div
        class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
      >
        <div
          class="MuiTypography-root MuiTypography-h5 MuiTypography-gutterBottom css-h93ljk-MuiTypography-root"
          data-testid="movie-card-title"
        >
          awesome movie
        </div>
        <p
          class="MuiTypography-root MuiTypography-body2 css-r40f8v-MuiTypography-root"
          data-testid="movie-card-description"
        >
          some overview
        </p>
        <hr />
        <p
          class="MuiTypography-root MuiTypography-body2 css-r40f8v-MuiTypography-root"
          data-testid="movie-card-popularity"
        >
          Popularity: 
          9999
        </p>
        <div
          class="MuiGrid-root MuiGrid-container css-1xzqqmo-MuiGrid-root"
        >
          <p
            class="MuiTypography-root MuiTypography-body2 css-r40f8v-MuiTypography-root"
            data-testid="movie-card-rate"
          >
            9.9
             (
            777
            )
          </p>
          <p
            class="MuiTypography-root MuiTypography-body2 MuiTypography-alignRight css-15l36n3-MuiTypography-root"
            data-testid="movie-card-date"
          >
            05-05-2022
          </p>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
